File: bright.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

const (
	PATH       = "/sys/class/backlight/intel_backlight/brightness"
	MIN        = 0
	MAX        = 7500
	PCTMIN     = 2
	PCTMAX     = 100
	STEP   int = MAX / 100
	STEPS      = 20
)

var (
	LOGMIN = math.Log10(PCTMIN)
	LOGMAX = math.Log10(PCTMAX)
)

// GetBrightness returns the brightness in %.
func GetBrightness() (int, error) {
	f, err := os.ReadFile(PATH)
	if err != nil {
		return -1, fmt.Errorf("failed to read brightness: %v", err)
	}
	n, err := strconv.Atoi(string(f)[:len(f)-1])
	if err != nil {
		return -1, fmt.Errorf("didn't get a number: %v", err)
	}
	return n, nil
}

func GetBrightnessPct() (float64, error) {
	i, err := GetBrightness()
	if err != nil {
		return -1, err
	}
	return (float64(i) / float64(MAX)) * 100.0, nil
}

func SetBrightness(val int) error {
	if val > MAX {
		return fmt.Errorf("value greater than max brightness %d", MAX)
	}
	s := strconv.Itoa(val)
	return os.WriteFile(PATH, []byte(s), 0664)
}

func SetBrightnessPct(pct float64) error {
	return SetBrightness(int((pct / 100) * MAX))
}

func SetBrightnessSmooth(val int) error {
	current, err := GetBrightness()
	if err != nil {
		return err
	}
	if val > current {
		for ; current < val; current += STEP {
			SetBrightness(current)
			time.Sleep(10 * time.Millisecond)
		}
	} else {
		for ; current > val; current -= STEP {
			SetBrightness(current)
			time.Sleep(10 * time.Millisecond)
		}
	}
	return nil
}

func BrightnessPctToStep(pct float64) float64 {
	return math.Round(math.Log10(pct) / (LOGMAX - LOGMIN) * STEPS)
}

func StepToBrightnessPct(step float64) float64 {
	x := step / STEPS * (LOGMAX - LOGMIN)
	return math.Round(math.Max(math.Min(math.Pow(10, x), PCTMAX), PCTMIN))
}

// Usage prints the program usage, then exits with error code 1.
func Usage() {
	fmt.Printf("usage\n\t%s -U / -B / -D\n", os.Args[0])
	os.Exit(1)
}

func main() {
	if len(os.Args) < 2 || !(os.Args[1] == "-U" || os.Args[1] == "-D" || os.Args[1] == "-B") {
		Usage()
	}

	currentBrightness, err := GetBrightnessPct()
	if err != nil {
		fmt.Println(err)
	}

	if currentBrightness == 0 {
		SetBrightnessPct(PCTMIN + 1)
		currentBrightness, err = GetBrightnessPct()
		if err != nil {
			fmt.Println(err)
		}
	}

	currentStep := BrightnessPctToStep(currentBrightness)

	newStep := currentStep

	switch os.Args[1] {
	case "-U":
		newStep += 2
	case "-D":
		newStep -= 2
	case "-B":
		newStep += 5
	}

	newBrightness := StepToBrightnessPct(newStep)

	fmt.Printf("Current backlight: %.1f\nChanging to %.1f\n", currentBrightness, newBrightness)

	if newBrightness == 0 {
		os.Exit(1)
	}
	if os.Args[1] == "-U" && currentBrightness == 99.0 {
		newBrightness = 100.0
	}
	err = SetBrightnessSmooth(int((newBrightness / 100.0) * MAX))
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

File: bright.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define PATH "BACKLIGHT_PATH"
#define MIN 0
#define MAX 7500
#define PCTMIN 2
#define PCTMAX 100
#define STEP MAX / 100
#define STEPS 20

float LOGMIN;
float LOGMAX;

int GetBrightness() {
  FILE *f = fopen(PATH, "r");
  if (f == NULL)
    return -1;
  int n;
  fscanf(f, "%i", &n);
  fclose(f);
  return n;
}

float GetBrightnessPct() {
  int b = GetBrightness();
  if (b == -1)
    return -1;
  return ((float)b / MAX) * 100;
}

int SetBrightness(int val) {
  if (val > MAX)
    return -1;
  FILE *f = fopen(PATH, "w");
  int e = fprintf(f, "%d", val);
  fclose(f);
  return e;
}

int SetBrightnessPct(float pct) { return SetBrightness((pct / 100) * MAX); }

int SetBrightnessSmooth(int val) {
  int current = GetBrightness();
  struct timespec wait;
  wait.tv_nsec = 1e7;
  if (current < 0)
    return current;
  if (val > current) {
    for (; current <= val; current += STEP) {
      SetBrightness(current);
      nanosleep(&wait, &wait);
    }
  } else {
    for (; current >= val; current -= STEP) {
      SetBrightness(current);
      nanosleep(&wait, &wait);
    }
  }
  return 0;
}

float BrightnessPctToStep(float pct) {
  return round(log10(pct) / (LOGMAX - LOGMIN) * STEPS);
}

float StepToBrightnessPct(float step) {
  float x = step / STEPS * (LOGMAX - LOGMIN);
  return round(fmax(fmin(pow(10, x), PCTMAX), PCTMIN));
}

void Usage(char *argv[]) {
  printf("usage\n\t%s -U / -B / -D\n", argv[0]);
  exit(1);
}

int main(int argc, char *argv[]) {
  LOGMIN = log10(PCTMIN);
  LOGMAX = log10(PCTMAX);

  if (argc < 2 || !(!strcmp(argv[1], "-U") || !strcmp(argv[1], "-D") ||
                    !strcmp(argv[1], "-B")))
    Usage(argv);

  float currentBrightness = GetBrightnessPct();
  if (currentBrightness < 0)
    printf("Failed to read file \"%s\": %i\n", PATH, (int)currentBrightness);

  if (currentBrightness == 0) {
    SetBrightnessPct(PCTMIN + 1);
    currentBrightness = GetBrightnessPct();
    if (currentBrightness < 0)
      printf("Failed to read file \"%s\": %i\n", PATH, (int)currentBrightness);
  }

  float currentStep = BrightnessPctToStep(currentBrightness);

  float newStep = currentStep;

  if (strcmp(argv[1], "-D") == 0)
    newStep -= 2;
  else if (strcmp(argv[1], "-U") == 0)
    newStep += 2;
  else if (strcmp(argv[1], "-B") == 0)
    newStep += 5;

  float newBrightness = StepToBrightnessPct(newStep);

  printf("Current backlight: %.1f\nChanging to %.1f\n", currentBrightness,
         newBrightness);

  if (newBrightness == 0)
    exit(1);

  if (strcmp(argv[1], "-U") && currentBrightness == 99)
    newBrightness = 100;

  int e = SetBrightnessSmooth((newBrightness / 100) * MAX);
  if (e < 0) {
    printf("Failed to write brightness: %i\n", e);
    exit(1);
  }
}
